include/dnscpp/lookup.h

```diff
@@ -95,6 +103,11 @@ class Lookup : public Operation
      *  @return bool        was a result reported to userspace?
      */
     virtual bool execute(double now) = 0;
+
+    /**
+     *  This lookup is done. Either it exhausted all of its attempts or it has a result. Invoke the callback handler
+     */
+    virtual void finalize() = 0;
```

Comment:
If you leave it up to the caller to decide when the lookup is over, the Lookup should not have methods like 'expired()', 'exhausted()' etc.